Add Datadog Cluster Agent integration

davidor opened the pull request:
### What does this PR do?
Adds an integration for the [Cluster Agent for Kubernetes](https://docs.datadoghq.com/agent/cluster_agent/).

### Additional Notes
This is very similar to a previous PR by @juliogreff (#7590) that was closed without merging. I started from a new integration to make sure to have the latest changes in the scaffolding code.

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] PR title must be written as a CHANGELOG entry [(see why)](https://github.com/DataDog/integrations-core/blob/master/CONTRIBUTING.md#pull-request-title)
- [ ] Files changes must correspond to the primary purpose of the PR as described in the title (small unrelated changes should have their own PR)
- [ ] PR must have `changelog/` and `integration/` labels attached

ahmed-mez commented on datadog_cluster_agent/datadog_checks/datadog_cluster_agent/check.py:
```diff
@@ -0,0 +1,56 @@
+                    'metrics': [DEFAULT_METRICS],
+                    'label_joins': {
+                        'leader_election_is_leader': {
+                            'labels_to_match': ['join_leader'],
```
We can leverage https://github.com/DataDog/integrations-core/pull/7769 to add the `is_leader` tag to every metric.
```suggestion
                            'labels_to_match': ['*'],
```

ahmed-mez reviewed (commented):
Nice work! Just one small suggestion - let's also add a `auto_conf.yaml` file

davidor commented:
Thanks for the review @ahmed-mez  I'll make those changes :+1:

We're discussing whether we should use `OpenMetricsBaseCheckV2` instead of `OpenMetricsBaseCheck` in this integration.

cswatt reviewed (approved):
looks good for docs

ChristineTChen commented on datadog_cluster_agent/assets/configuration/spec.yaml:
```diff
@@ -0,0 +1,12 @@
+  options:
+  - template: init_config
+    options:
+      - template: init_config/openmetrics
```
```suggestion
      - template: init_config/openmetrics_legacy
```
Needs resync

`init_config/openmetrics` template refers to options that correspond with OpenmetricsV2BaseCheck

davidor replied:
You're right, unfortunately I realized that I was using the wrong template after submitting the PR and seeing that the config github check failed.

@clamoriniere and @hithwen proposed to make this integration work with `OpenMetricsV2BaseCheck` and fallback to `OpenMetricsBaseCheck` when running on Python 2 (V2 only works with python 3). The idea is to implement something similar to what you already have in the linkerd integration.

Do you think we should merge this one first with `OpenMetricsBaseCheck` and `init_config/openmetrics_legacy` and then do a separate PR with V2 and the fallback or should I just try to do everything in this one?

ChristineTChen commented on datadog_cluster_agent/assets/configuration/spec.yaml:
```diff
@@ -0,0 +1,12 @@
+      - template: init_config/openmetrics
+  - template: instances
+    options:
+      - template: instances/openmetrics
```
```suggestion
      - template: instances/openmetrics_legacy
```
same here

ChristineTChen commented on datadog_cluster_agent/manifest.json:
```diff
@@ -0,0 +1,36 @@
+  "manifest_version": "1.0.0",
+  "name": "datadog_cluster_agent",
+  "metric_prefix": "datadog_cluster_agent.",
+  "metric_to_check": "",
```
need a metric here
```suggestion
  "metric_to_check": "datadog.cluster_agent.admission_webhooks.certificate_expiry",
```

ChristineTChen commented on datadog_cluster_agent/manifest.json:
```diff
@@ -0,0 +1,36 @@
+    "log collection"
+  ],
+  "type": "check",
+  "is_public": false,
```
```suggestion
  "is_public": true,
```

ChristineTChen reviewed (changes requested):
Looks good! Left some comments to correct the options in conf.yaml.example

davidor commented:
@ChristineTChen , @ahmed-mez thanks for the reviews. I addressed all your comments.

Regarding `OpenMetricsV2BaseCheck` vs `OpenMetricsBaseCheck`, V2 changes a bit the behavior around `join_labels` and it affects this specific use case. Once there's an option to configure the same behavior as V1, we'll open a new PR with V2 plus a fallback to V1 when using python 2.

ChristineTChen commented on datadog_cluster_agent/assets/configuration/spec.yaml:
```diff
@@ -0,0 +1,12 @@
+    options:
+      - template: instances/openmetrics_legacy
+        overrides:
+          prometheus_url.value.example: http://localhost:5000/metrics
```
```suggestion
          prometheus_url.value.example: http://localhost:5000/metrics

- name: auto_conf.yaml
  options:
  - template: ad_identifiers
    overrides:
      value.example:
        - cluster-agent
  - template: init_config
    options: []
  - template: instances
    options:
    - name: prometheus_url
      description: The URL where your application metrics are exposed by Prometheus.
      value:
        example: http://%%host%%:5000/metrics
        type: string
```

You can include auto_conf.yaml here as well

davidor replied:
Thanks. I added it and regenerated the sample file.

ChristineTChen reviewed (approved):
Thanks!

ahmed-mez commented on datadog_cluster_agent/datadog_checks/datadog_cluster_agent/data/auto_conf.yaml:
```diff
@@ -0,0 +1,20 @@
+
+## Every instance is scheduled independent of the others.
+#
+instances:
+
+    ## @param prometheus_url - string - required
+    ## The URL where your application metrics are exposed by Prometheus.
+    #
+  - prometheus_url: http://%%host%%:5000/metrics
```
Let's also add the `is_leader` label joins config to this config, wdyt?

ahmed-mez replied:
Confirmed that the default config will still apply

davidor replied:
Yes, I also added the changes in the spec.yaml so the ad template has the same options as the default one.

ahmed-mez reviewed (approved):
:shipit:
